### ISSUE TITLE
Views being scanned very often.  Too often?

### ISSUE THREAD

**Issue by kwerle** [rails-sqlserver/activerecord-sqlserver-adapter#552]
Yay!  Rails 5 support!  

The only "issue" I've noticed so far is that one of the views I use gets scanned an awful lot in debug mode:

  CACHE (0.0ms)  SELECT * FROM INFORMATION_SCHEMA.VIEWS WHERE TABLE_NAME = 'vAddress_Extn'

Maybe this is just debug info, or maybe it isn't being cached quite the same way as table information.  In any case, it does not seem to be doing any harm - but I thought I'd mention it because it's not what I expect.

Thank you!

**Comment by metaskills** [comment]
Thanks. and good eye - you are right on both fronts. First, we removed the custom SchemaCache we had in lower Rails version. This was getting harder to maintain and none of the core adapters  use a custom schema cache anymore. When I got close to finishing the Rails v5 work, I did an audit of methods that would call for views to make sure we only call these methods when reflecting on the schema. See these two commits:

* https://github.com/rails-sqlserver/activerecord-sqlserver-adapter/commit/237ce43728c39e11c9a166bd1f31b9ae5cb5a743
* https://github.com/rails-sqlserver/activerecord-sqlserver-adapter/commit/611e81af037f1e9db4633fb3f50b06eb24dc8503

One neat feature of Rails that I really want to make sure our adapter supports well is the `db:schema:cache:dump` task. See here for full information:

**Rails feature that you've never heard about: schema cache**
http://blog.iempire.ru/2016/12/13/schema-cache/

By removing the our custom caching, we should be able to support this feature seamlessly. I also did a quick search into the ActiveRecord source for 5-0-stable and it appears using this feature is not limited to production too. Can you test to see if dumping the schema cache assets you in development? Would be great to find out.

**Comment by kwerle** [comment]
After nearly an hour of running!  (with no logging/output) I got the following:

ActiveRecord::StatementInvalid: TinyTds::Error: Incorrect syntax near 'InstallationID'.: SELECT c.COLUMN_DEFAULT FROM .INFORMATION_SCHEMA.COLUMNS c WHERE c.TABLE_NAME = ''' AND c.COLUMN_NAME = 'InstallationID'

There are about 227 references to columns named InstallationID in my schema.
There are around 110 references to indexes that include InstallationID in my schema.
For reference, my schema.rb file is > 12K lines.

**Comment by metaskills** [comment]
So the SQL for that error should be a lot longer and include the remaining SQL which is defined in the column_definitions method. Also, did you do a schema cache dump? Just need more context to help.

**Comment by kwerle** [comment]
I'm getting 2-4 times slower performance in Rails5 for some of my pages.  I just don't have the bandwidth to deal with that at the moment.

I'll try to get you some error output - but at 1 hour/shot, it's slow going.  I should be able to set up a container and ignore it, tomorrow, though.

**Comment by metaskills** [comment]
I can help. I do think there is something not right. Technically you should not see that views reflection in production.

**Comment by kwerle** [comment]
This is about all I'll be able to provide:

```
root@f41878c43d2e:/usr/src/app# rake db:schema:cache:dump
/usr/local/bundle/gems/activesupport-5.0.1/lib/active_support/xml_mini.rb:51: warning: constant ::Fixnum is deprecated
/usr/local/bundle/gems/activesupport-5.0.1/lib/active_support/xml_mini.rb:52: warning: constant ::Bignum is deprecated
/usr/local/bundle/gems/sprockets-3.3.5/lib/sprockets/processor_utils.rb:110: warning: constant ::Fixnum is deprecated
/usr/local/bundle/gems/sprockets-3.3.5/lib/sprockets/processor_utils.rb:111: warning: constant ::Bignum is deprecated
/usr/local/bundle/gems/activesupport-5.0.1/lib/active_support/core_ext/numeric/conversions.rb:138: warning: constant ::Fixnum is deprecated
[DEPRECATION] the dotenv-deployment gem is deprecated. See https://github.com/bkeepers/dotenv-deployment#readme.
  SQL (47.1ms)  USE [SolarWorks]
rake aborted!
ActiveRecord::StatementInvalid: TinyTds::Error: Incorrect syntax near 'InstallationID'.: SELECT c.COLUMN_DEFAULT FROM .INFORMATION_SCHEMA.COLUMNS c WHERE c.TABLE_NAME = ''' AND c.COLUMN_NAME = 'InstallationID'
/usr/local/bundle/gems/activerecord-sqlserver-adapter-5.0.1/lib/active_record/connection_adapters/sqlserver/database_statements.rb:320:in `each'
/usr/local/bundle/gems/activerecord-sqlserver-adapter-5.0.1/lib/active_record/connection_adapters/sqlserver/database_statements.rb:320:in `handle_to_names_and_values_dblib'
/usr/local/bundle/gems/activerecord-sqlserver-adapter-5.0.1/lib/active_record/connection_adapters/sqlserver/database_statements.rb:311:in `handle_to_names_and_values'
/usr/local/bundle/gems/activerecord-sqlserver-adapter-5.0.1/lib/active_record/connection_adapters/sqlserver/database_statements.rb:290:in `_raw_select'
/usr/local/bundle/gems/activerecord-sqlserver-adapter-5.0.1/lib/active_record/connection_adapters/sqlserver/database_statements.rb:285:in `block in raw_select'
/usr/local/bundle/gems/activerecord-5.0.1/lib/active_record/connection_adapters/abstract_adapter.rb:589:in `block in log'
/usr/local/bundle/gems/activesupport-5.0.1/lib/active_support/notifications/instrumenter.rb:21:in `instrument'
/usr/local/bundle/gems/activerecord-5.0.1/lib/active_record/connection_adapters/abstract_adapter.rb:583:in `log'
/usr/local/bundle/gems/activerecord-sqlserver-adapter-5.0.1/lib/active_record/connection_adapters/sqlserver/database_statements.rb:285:in `raw_select'
/usr/local/bundle/gems/activerecord-sqlserver-adapter-5.0.1/lib/active_record/connection_adapters/sqlserver/database_statements.rb:226:in `sp_executesql'
/usr/local/bundle/gems/activerecord-sqlserver-adapter-5.0.1/lib/active_record/connection_adapters/sqlserver/database_statements.rb:7:in `select_rows'
/usr/local/bundle/gems/activerecord-5.0.1/lib/active_record/connection_adapters/abstract/database_statements.rb:54:in `select_value'
/usr/local/bundle/gems/activerecord-sqlserver-adapter-5.0.1/lib/active_record/connection_adapters/sqlserver/schema_statements.rb:379:in `block in column_definitions'
/usr/local/bundle/gems/activerecord-5.0.1/lib/active_record/result.rb:52:in `block in each'
/usr/local/bundle/gems/activerecord-5.0.1/lib/active_record/result.rb:52:in `each'
/usr/local/bundle/gems/activerecord-5.0.1/lib/active_record/result.rb:52:in `each'
/usr/local/bundle/gems/activerecord-sqlserver-adapter-5.0.1/lib/active_record/connection_adapters/sqlserver/schema_statements.rb:355:in `map'
/usr/local/bundle/gems/activerecord-sqlserver-adapter-5.0.1/lib/active_record/connection_adapters/sqlserver/schema_statements.rb:355:in `column_definitions'
/usr/local/bundle/gems/activerecord-sqlserver-adapter-5.0.1/lib/active_record/connection_adapters/sqlserver/schema_statements.rb:75:in `columns'
/usr/local/bundle/gems/activerecord-5.0.1/lib/active_record/connection_adapters/schema_cache.rb:56:in `columns'
/usr/local/bundle/gems/activerecord-sqlserver-adapter-5.0.1/lib/active_record/connection_adapters/sqlserver/schema_statements.rb:106:in `primary_keys'
/usr/local/bundle/gems/activerecord-5.0.1/lib/active_record/connection_adapters/abstract/schema_statements.rb:132:in `primary_key'
/usr/local/bundle/gems/activerecord-5.0.1/lib/active_record/connection_adapters/schema_cache.rb:25:in `primary_keys'
/usr/local/bundle/gems/activerecord-5.0.1/lib/active_record/connection_adapters/schema_cache.rb:42:in `add'
/usr/local/bundle/gems/activerecord-5.0.1/lib/active_record/railties/databases.rake:274:in `block (5 levels) in <top (required)>'
/usr/local/bundle/gems/activerecord-5.0.1/lib/active_record/railties/databases.rake:274:in `each'
/usr/local/bundle/gems/activerecord-5.0.1/lib/active_record/railties/databases.rake:274:in `block (4 levels) in <top (required)>'
/usr/local/bundle/gems/rake-11.3.0/exe/rake:27:in `<top (required)>'
TinyTds::Error: Incorrect syntax near 'InstallationID'.
/usr/local/bundle/gems/activerecord-sqlserver-adapter-5.0.1/lib/active_record/connection_adapters/sqlserver/database_statements.rb:320:in `each'
/usr/local/bundle/gems/activerecord-sqlserver-adapter-5.0.1/lib/active_record/connection_adapters/sqlserver/database_statements.rb:320:in `handle_to_names_and_values_dblib'
/usr/local/bundle/gems/activerecord-sqlserver-adapter-5.0.1/lib/active_record/connection_adapters/sqlserver/database_statements.rb:311:in `handle_to_names_and_values'
/usr/local/bundle/gems/activerecord-sqlserver-adapter-5.0.1/lib/active_record/connection_adapters/sqlserver/database_statements.rb:290:in `_raw_select'
/usr/local/bundle/gems/activerecord-sqlserver-adapter-5.0.1/lib/active_record/connection_adapters/sqlserver/database_statements.rb:285:in `block in raw_select'
/usr/local/bundle/gems/activerecord-5.0.1/lib/active_record/connection_adapters/abstract_adapter.rb:589:in `block in log'
/usr/local/bundle/gems/activesupport-5.0.1/lib/active_support/notifications/instrumenter.rb:21:in `instrument'
/usr/local/bundle/gems/activerecord-5.0.1/lib/active_record/connection_adapters/abstract_adapter.rb:583:in `log'
/usr/local/bundle/gems/activerecord-sqlserver-adapter-5.0.1/lib/active_record/connection_adapters/sqlserver/database_statements.rb:285:in `raw_select'
/usr/local/bundle/gems/activerecord-sqlserver-adapter-5.0.1/lib/active_record/connection_adapters/sqlserver/database_statements.rb:226:in `sp_executesql'
/usr/local/bundle/gems/activerecord-sqlserver-adapter-5.0.1/lib/active_record/connection_adapters/sqlserver/database_statements.rb:7:in `select_rows'
/usr/local/bundle/gems/activerecord-5.0.1/lib/active_record/connection_adapters/abstract/database_statements.rb:54:in `select_value'
/usr/local/bundle/gems/activerecord-sqlserver-adapter-5.0.1/lib/active_record/connection_adapters/sqlserver/schema_statements.rb:379:in `block in column_definitions'
/usr/local/bundle/gems/activerecord-5.0.1/lib/active_record/result.rb:52:in `block in each'
/usr/local/bundle/gems/activerecord-5.0.1/lib/active_record/result.rb:52:in `each'
/usr/local/bundle/gems/activerecord-5.0.1/lib/active_record/result.rb:52:in `each'
/usr/local/bundle/gems/activerecord-sqlserver-adapter-5.0.1/lib/active_record/connection_adapters/sqlserver/schema_statements.rb:355:in `map'
/usr/local/bundle/gems/activerecord-sqlserver-adapter-5.0.1/lib/active_record/connection_adapters/sqlserver/schema_statements.rb:355:in `column_definitions'
/usr/local/bundle/gems/activerecord-sqlserver-adapter-5.0.1/lib/active_record/connection_adapters/sqlserver/schema_statements.rb:75:in `columns'
/usr/local/bundle/gems/activerecord-5.0.1/lib/active_record/connection_adapters/schema_cache.rb:56:in `columns'
/usr/local/bundle/gems/activerecord-sqlserver-adapter-5.0.1/lib/active_record/connection_adapters/sqlserver/schema_statements.rb:106:in `primary_keys'
/usr/local/bundle/gems/activerecord-5.0.1/lib/active_record/connection_adapters/abstract/schema_statements.rb:132:in `primary_key'
/usr/local/bundle/gems/activerecord-5.0.1/lib/active_record/connection_adapters/schema_cache.rb:25:in `primary_keys'
/usr/local/bundle/gems/activerecord-5.0.1/lib/active_record/connection_adapters/schema_cache.rb:42:in `add'
/usr/local/bundle/gems/activerecord-5.0.1/lib/active_record/railties/databases.rake:274:in `block (5 levels) in <top (required)>'
/usr/local/bundle/gems/activerecord-5.0.1/lib/active_record/railties/databases.rake:274:in `each'
/usr/local/bundle/gems/activerecord-5.0.1/lib/active_record/railties/databases.rake:274:in `block (4 levels) in <top (required)>'
/usr/local/bundle/gems/rake-11.3.0/exe/rake:27:in `<top (required)>'
Tasks: TOP => db:schema:cache:dump
(See full trace by running task with --trace)

```

**Comment by metaskills** [comment]
> awful lot in debug mode:
> CACHE (0.0ms) SELECT * FROM INFORMATION_SCHEMA.VIEWS WHERE TABLE_NAME ...

I am trying to debug how you see that log line. Can you tell me what debug mode means? I am having a hard time reproducing this line showing up in the logs. I can force it by changing some SQL names from `SCHEMA` to `SQL` in the `view_information` format... and I see where that code should fire when any view based model is first hit. But it should never be visible in the log. Much less, via a query cache. I'd like to know more about how you got that log line? TIA.

**Comment by metaskills** [comment]
OK, I looked into this a bit more. I even used a simple Rails 5 smoke test app that had one table and one view. Here is where we are:

* View information been called for schema is normal. This will happen once for each column in each table. Essentially `t x c` or `tables x columns`.
* This is expensive, but should be cached both in development, test, production. 

### Simple Test App

Use this to help you deduce why your setup is different.

https://github.com/rails-sqlserver/ss_app

### @kwerle Questions

* Find out why your env is not caching these. Maybe you have eager load on or something?
* Find out why your log is showing private SCHEMA entries?

### My Actions

1. I think I can look into optimizing schema reflection and only call view info once per table vs once per column per table.
2. Make the schema cache dump work for you. I see you have a nil `database` value in that method. I'll see if I can reproduce how that happened.

**Comment by metaskills** [comment]
@kwerle WRT #2, can you do the following:

```shell
$ bundle open activerecord-sqlserver-adapter
```

Find the `schema_statements.rb` file and then the `def column_definitions(table_name)` method around line 293. Put a `ActiveRecord.logger.debug` or `puts` statement of the `table_name` value. Then run the db schema dump task again. 

This will give you the name of the table given to this method before it crashes. I am very interested in what that table name is and why `database    = identifier.fully_qualified_database_quoted` might be nil.

**Comment by metaskills** [comment]
Math update, the way the code is written, `view_information` is called `views x column` vs `tables`

**Comment by metaskills** [comment]
I made a small change that will be released in v5.0.3 later today that memoizes view information to per table. I left some debug details for the issue of the schema cache dump. But that feels like a distinct issue maybe we can hash out in another issue.
